### PR TITLE
dbengine free from RRDSET and RRDDIM

### PR DESCRIPTION
**Pull request by ktsaou**
- [x] RRDDIM and RRDSET are no longer used by the dbengine API. This unblocks #13697 
- [x] dbengine now ensures that pages are written in a consistent manner and reports inconsistencies when loading journal files.

### CODE REVIEW

**Comment by ktsaou** [comment]
This is ready for review. I depend on this for #13697 which must be finished asap.
So, please have a look.

File: database/engine/rrdengineapi.c [hunk 1]
```diff
@@ -251,50 +308,66 @@ void rrdeng_store_metric_next(STORAGE_COLLECT_HANDLE *collection_handle,
...
     void *page;
     uint8_t must_flush_unaligned_page = 0, perfect_page_alignment = 0;
 
     if (descr) {
         /* Make alignment decisions */
 
-        if (descr->page_length == rd->rrdset->rrddim_page_alignment) {
+#ifdef NETDATA_INTERNAL_CHECKS
+        if(descr->end_time_ut + page_index->latest_update_every_s * USEC_PER_SEC != point_in_time_ut) {
+            char buffer[200];
```
Comment (thiagoftsm) [inline database/engine/rrdengineapi.c:324]: A minor suggestion. In normal situation in our code we set arrays like this with a size bigger than we give to `snprintfz`, for example, this [json](https://github.com/netdata/netdata/blob/294390f3b7514e8e23f3eb840be7eb25de442d6f/web/api/formatters/json/json.c#L20-L56) code, but here you are using the same size. I suggest to keep pattern and set here size `201`.

File: database/ram/rrddim_mem.c [hunk 2]
```diff
@@ -2,18 +2,67 @@
...
 
-STORAGE_METRIC_HANDLE *rrddim_metric_init(RRDDIM *rd, STORAGE_INSTANCE *db_instance __maybe_unused) {
+STORAGE_METRIC_HANDLE *
+rrddim_metric_get_or_create(RRDDIM *rd, STORAGE_INSTANCE *db_instance __maybe_unused, STORAGE_METRICS_GROUP *smg __maybe_unused) {
+    STORAGE_METRIC_HANDLE *t = rrddim_metric_get(db_instance, &rd->metric_uuid, smg);
+    if(!t) {
+        netdata_rwlock_wrlock(&rrddim_JudyHS_rwlock);
+        Pvoid_t *PValue = JudyHSIns(&rrddim_JudyHS_array, &rd->metric_uuid, sizeof(uuid_t), PJE0);
+        fatal_assert(NULL == *PValue);
+        *PValue = rd;
```
Comment (thiagoftsm) [inline database/ram/rrddim_mem.c:31]: Like previous PRs, I ran coverity on this PR and we got few warnings, like this:

```sh
database/ram/rrddim_mem.c: 34 in rrddim_metric_get_or_create()
28             netdata_rwlock_wrlock(&rrddim_JudyHS_rwlock);
29             Pvoid_t *PValue = JudyHSIns(&rrddim_JudyHS_array, &rd->metric_uuid, sizeof(uuid_t), PJE0);
30             fatal_assert(NULL == *PValue);
31             *PValue = rd;
32             t = (STORAGE_METRIC_HANDLE *)rd;
33             netdata_rwlock_unlock(&rrddim_JudyHS_rwlock);
>>>     CID 358386:  Resource leaks  (RESOURCE_LEAK)
>>>     Variable "PValue" going out of scope leaks the storage it points to.
34         }
35     
36         if((RRDDIM *)t != rd)
37             fatal("RRDDIM_MEM: incorrect pointer returned from index.");
38     
39         return (STORAGE_METRIC_HANDLE *)rd;
```

File: database/engine/rrdengine.c [hunk 3]
```diff
@@ -1043,7 +1043,70 @@ static void rrdeng_cleanup_finished_threads(struct rrdengine_worker_config* wc)
...
 {
-    return init_data_files(ctx);
+    int ret = init_data_files(ctx);
+
+    BUFFER *wb = buffer_create(1000);
+    size_t all_errors = 0;
+    usec_t now = now_realtime_usec();
+
+    if(ctx->load_errors[LOAD_ERRORS_PAGE_FLIPPED_TIME].counter) {
+        buffer_sprintf(wb, "%s%zu pages had start time > end time (latest: %llu secs ago)"
```
Comment (thiagoftsm) [inline database/engine/rrdengine.c:1053]: We also have this:

```sh
/database/engine/rrdengine.c: 1053 in init_rrd_files()
1047     
1048         BUFFER *wb = buffer_create(1000);
1049         size_t all_errors = 0;
1050         usec_t now = now_realtime_usec();
1051     
1052         if(ctx->load_errors[LOAD_ERRORS_PAGE_FLIPPED_TIME].counter) {
>>>     CID 358385:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach the expression "", "" inside this statement: "buffer_sprintf(wb, "%s%zu p...".
1053             buffer_sprintf(wb, "%s%zu pages had start time > end time (latest: %llu secs ago)"
1054                            , (all_errors)?", ":""
1055                            , ctx->load_errors[LOAD_ERRORS_PAGE_FLIPPED_TIME].counter
1056                            , (now - ctx->load_errors[LOAD_ERRORS_PAGE_FLIPPED_TIME].latest_end_time_ut) / USEC_PER_SEC
1057                            );
1058             all_errors += ctx->load_errors[LOAD_ERRORS_PAGE_FLIPPED_TIME].counter;
```

File: database/engine/rrdengineapi.c [hunk 4]
```diff
@@ -75,97 +103,128 @@ void rrdeng_convert_legacy_uuid_to_multihost(char machine_guid[GUID_LEN + 1], uu
...
+    struct rrdengine_instance *ctx = (struct rrdengine_instance *)db_instance;
+    struct pg_alignment *pa = (struct pg_alignment *)smg;
+    struct pg_cache_page_index *page_index;
+    struct page_cache *pg_cache = &ctx->pg_cache;
 
-        int need_to_store = uuid_compare(rd->metric_uuid, multihost_legacy_uuid);
+    uv_rwlock_wrlock(&pg_cache->metrics_index.lock);
+    Pvoid_t *PValue = JudyHSIns(&pg_cache->metrics_index.JudyHS_array, uuid, sizeof(uuid_t), PJE0);
+    fatal_assert(NULL == *PValue); /* TODO: figure out concurrency model */
+    *PValue = page_index = create_page_index(uuid);
```
Comment (thiagoftsm) [inline database/engine/rrdengineapi.c:173]: We also have this:

```sh
176         pg_cache->metrics_index.last_page_index = page_index;
177         page_index->alignment = pa;
178         if(pa)
179             pa->refcount++;
180         uv_rwlock_wrunlock(&pg_cache->metrics_index.lock);
181     
>>>     CID 358387:  Resource leaks  (RESOURCE_LEAK)
>>>     Variable "PValue" going out of scope leaks the storage it points to.
182         return (STORAGE_METRIC_HANDLE *)page_index;
183     }
184     
185     STORAGE_METRIC_HANDLE *rrdeng_metric_get_or_create(RRDDIM *rd, STORAGE_INSTANCE *db_instance, STORAGE_METRICS_GROUP *smg) {
186         STORAGE_METRIC_HANDLE *db_metric_handle;
187     
```

I think we already talked about this in previous PR, but I am bringing these leaks that Judy adds to bring a complete report.

File: database/engine/journalfile.c [hunk 5]
```diff
@@ -348,21 +374,30 @@ static void restore_extent_metadata(struct rrdengine_instance *ctx, struct rrden
...
         descr->extent = extent;
         descr->type = page_type;
         extent->pages[valid_pages++] = descr;
         pg_cache_insert(ctx, page_index, descr);
+
+        if(page_index->latest_time_ut == descr->end_time_ut)
+            page_index->latest_update_every_s = descr->update_every_s;
+
+        if(descr->update_every_s == 0)
+            fatal("DBENGINE: page descriptor update every is zero, end_time_ut = %lu, start_time_ut = %lu, entries = %zu", end_time_ut, start_time_ut, entries);
```
Comment (thiagoftsm) [inline database/engine/journalfile.c:390]: On Armvl7 (32 bits) this line inserts these warnings:

```sh
database/engine/journalfile.c: In function 'restore_extent_metadata':
database/engine/journalfile.c:379:56: warning: operand of '?:' changes signedness from 'time_t' {aka 'long int'} to 'uint32_t' {aka 'unsigned int'} due to unsignedness of other operand [-Wsign-compare]
  379 |         descr->update_every_s = (update_every_s > 0) ? update_every_s : (page_index->latest_update_every_s);
      |                                                        ^~~~~~~~~~~~~~
In file included from ./libnetdata/libnetdata.h:420,
                 from ./daemon/common.h:6,
                 from database/engine/rrdengine.h:14,
                 from database/engine/journalfile.c:2:
database/engine/journalfile.c:390:19: warning: format '%lu' expects argument of type 'long unsigned int', but argument 5 has type 'uint64_t' {aka 'long long unsigned int'} [-Wformat=]
  390 |             fatal("DBENGINE: page descriptor update every is zero, end_time_ut = %lu, start_time_ut = %lu, entries = %zu", end_time_ut, start_time_ut, entries);
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~
      |                                                                                                                            |
      |                                                                                                                            uint64_t {aka long long unsigned int}
./libnetdata/log/log.h:98:72: note: in definition of macro 'fatal'
   98 | #define fatal(args...)   fatal_int(__FILE__, __FUNCTION__, __LINE__, ##args)
      |                                                                        ^~~~
database/engine/journalfile.c:390:84: note: format string is defined here
  390 |             fatal("DBENGINE: page descriptor update every is zero, end_time_ut = %lu, start_time_ut = %lu, entries = %zu", end_time_ut, start_time_ut, entries);
      |                                                                                  ~~^
      |                                                                                    |
      |                                                                                    long unsigned int
      |                                                                                  %llu
In file included from ./libnetdata/libnetdata.h:420,
                 from ./daemon/common.h:6,
                 from database/engine/rrdengine.h:14,
                 from database/engine/journalfile.c:2:
database/engine/journalfile.c:390:19: warning: format '%lu' expects argument of type 'long unsigned int', but argument 6 has type 'uint64_t' {aka 'long long unsigned int'} [-Wformat=]
  390 |             fatal("DBENGINE: page descriptor update every is zero, end_time_ut = %lu, start_time_ut = %lu, entries = %zu", end_time_ut, start_time_ut, entries);
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~               ~~~~~~~~~~~~~
      |                                                                                                                                         |
      |                                                                                                                                         uint64_t {aka long long unsigned int}
./libnetdata/log/log.h:98:72: note: in definition of macro 'fatal'
   98 | #define fatal(args...)   fatal_int(__FILE__, __FUNCTION__, __LINE__, ##args)
      |                                                                        ^~~~
database/engine/journalfile.c:390:105: note: format string is defined here
  390 |             fatal("DBENGINE: page descriptor update every is zero, end_time_ut = %lu, start_time_ut = %lu, entries = %zu", end_time_ut, start_time_ut, entries);
      |                                                                                                       ~~^
      |                                                                                                         |
      |                                                                                                         long unsigned int
      |                                                                                                       %llu

```

**Comment by thiagoftsm** [comment]
Hello @ktsaou ,

This branch is running fine (Streaming, plugins, and alerts) on `Slackware current`, `CentOS 7.9`, `FreeBSD` and `Raspberry OS`, please, take a look on the  warnings we have. 

Best regards!

File: database/engine/rrdengine.c (hunk 3 above)
Comment (ktsaou) [reply database/engine/rrdengine.c:1053]: uniformity for the code below this.

File: database/ram/rrddim_mem.c (hunk 2 above)
Comment (ktsaou) [reply database/ram/rrddim_mem.c:31]: this is how judy works

File: database/engine/rrdengineapi.c (hunk 4 above)
Comment (ktsaou) [reply database/engine/rrdengineapi.c:173]: this is how judy works

File: database/engine/rrdengineapi.c (hunk 1 above)
Comment (ktsaou) [reply database/engine/rrdengineapi.c:324]: done

Comment (thiagoftsm) [reply database/engine/rrdengineapi.c:324]: I am compiling your branch right now and I will run a last round of tests. :handshake:

**Comment by ktsaou** [comment]
Done all changes requested.
Please review this. I need it merged to continue working on the query engine.

**Comment by ktsaou** [comment]
@thiagoftsm disable internal checks when you compile with asan. The problem is that the memory allocator of internal checks does a heap use outside of any allocation because aclk is mixing malloc/mallocz and free/freez. Until this is fixed, to use asan, disable internal checks and it should work.

**Comment by ktsaou** [comment]
@thiagoftsm could you please test memory modes other than dbengine?

**Comment by thiagoftsm** [comment]
> @thiagoftsm could you please test memory modes other than dbengine?

Of course, right now I am testing `parent` with `ram` and `children` with `dbengine`. I will test other too. :handshake:

**Review by thiagoftsm — DISMISSED** [review]
I ran this PR during hours using the following scenario:

| Host | Distribution |memory mode | Is parent? |
|--------|-----------------|---------------|-------|
| X86_64 | Slackware current | ram | yes |
| X86_64 libvirt VM| Arch Linux | map | no | 
| Armvl7 32 bits|  Raspberry OS | dbengine | no |
| X86_64 libvirt VM | CentOS 7.9 | none | no |

All tests worked as expected. I had streaming, collectors, alerts (on Arch), and cloud. LGTM!

**Comment by thiagoftsm** [comment]
Hello @vlvkobal ,

When you have a time, please, take a look at `cmocka`, because this is not the unique PR that I have seen issues with it.

Best regards!

File: database/engine/journalfile.c [hunk 6]
```diff
@@ -340,29 +366,40 @@ static void restore_extent_metadata(struct rrdengine_instance *ctx, struct rrden
...
             uv_rwlock_wrunlock(&pg_cache->metrics_index.lock);
         }
 
         descr = pg_cache_create_descr();
         descr->page_length = jf_metric_data->descr[i].page_length;
-        descr->start_time = start_time;
-        descr->end_time = end_time;
+        descr->start_time_ut = start_time_ut;
+        descr->end_time_ut = end_time_ut;
+        descr->update_every_s = (update_every_s > 0) ? (uint32_t)update_every_s : (page_index->latest_update_every_s);
```
Comment (thiagoftsm) [inline database/engine/journalfile.c:379]: I ran coverity scan again on this branch and now we have this:

```sh
*** CID 358453:  High impact quality  (Y2K38_SAFETY)
/database/engine/journalfile.c: 379 in restore_extent_metadata()
373             }
374     
375             descr = pg_cache_create_descr();
376             descr->page_length = jf_metric_data->descr[i].page_length;
377             descr->start_time_ut = start_time_ut;
378             descr->end_time_ut = end_time_ut;
>>>     CID 358453:  High impact quality  (Y2K38_SAFETY)
>>>     A "time_t" value is stored in an integer with too few bits to accommodate it.  The expression "update_every_s" is cast to "uint32_t".
379             descr->update_every_s = (update_every_s > 0) ? (uint32_t)update_every_s : (page_index->latest_update_every_s);
380             descr->id = &page_index->id;
381             descr->extent = extent;
382             descr->type = page_type;
383             extent->pages[valid_pages++] = descr;
384             pg_cache_insert(ctx, page_index, descr);
```

Comment (stelfrag) [reply database/engine/journalfile.c:379]: We can ignore this. It just stores the "update_every". Eventually we will need to deal with all the `Y2K38_SAFETY` warnings

**Review by stelfrag — APPROVED** [review]

**Review by thiagoftsm — APPROVED** [review]
I retest the PR on `x86_64` and `armvl7` and everything is working as expected.

**Pull request merged**